### ISSUE TITLE
[HOLD for payment 2022-05-20][$2000] Copying / pasting code between < and /> seems broken

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7912]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Type something like <Component/> and send the message
2. Try to select, copy, then paste into a new message

## Expected Result:
Exactly what you copied should be pasted

## Actual Result:
The code gets stripped
Issue happens when selecting with a mouse, not when using the Copy button

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.40-0
**Reproducible in staging?:**  Y 
**Reproducible in production?:** Y 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 

https://user-images.githubusercontent.com/43995119/155665755-5ae5607a-9621-4cb0-bf88-3606dd32f313.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @Beamanator 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1644856333330159

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @iwiznia (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @adelekennedy (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mateusbra** [comment]
### PROPOSAL
We have to change how we strip html when copying to clipboard, at the moment we are only replacing everything within <> for an empty string here:
https://github.com/Expensify/expensify-common/blob/f77bb4710c13d01153716df7fb087b637ba3b8bd/lib/str.js#L306-L318

**Comment by adelekennedy** [comment]
[internal](https://www.upwork.com/ab/applicants/1497360716363595776/job-details)
[external](https://www.upwork.com/jobs/~01c90fb0c06dde1e5c)

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @puneetlath (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by parasharrajat** [comment]
Thanks for the proposal @mateusbra. Could you please root cause details to your proposal? and tell us how did you reach your solution.

**Comment by dennismunene** [comment]
### **PROPOSAL**

We need to change how we handle pasted HTML here. 


**Update** https://github.com/Expensify/App/blob/36c6f8a2ccb4a846daff664036571c2cdf1fc164/src/components/TextInputFocusable/index.js#L258


To look like this. 

- Instead of getting the pasted from the clipboard content as HTML and converting to markdown ( which strips off all tags. see https://github.com/Expensify/expensify-common/blob/f77bb4710c13d01153716df7fb087b637ba3b8bd/lib/ExpensiMark.js#L373)  

- pick pasted content from clipboard as `Text`. This will stick have some html tags

```  
handlePaste(event) {
        const {files, types} = event.clipboardData;
        const TEXT_HTML = 'text/html';
        const TEXT = 'Text';

        // If paste contains files, then trigger file management
        if (files.length > 0) {
            // Prevent the default so we do not post the file name into the text box
            event.preventDefault();
            this.props.onPasteFile(event.clipboardData.files[0]);
            return;
        }

        // If paste contains HTML
        if (types.includes(TEXT_HTML)) {
            const pastedHTML = event.clipboardData.getData(TEXT_HTML);
            const pastedHTMLAsText = event.clipboardData.getData(TEXT);

            event.preventDefault();
            const domparser = new DOMParser();
            const embeddedImages = domparser.parseFromString(pastedHTML, TEXT_HTML)
                .images;

            // If HTML has img tag, then fetch images from it.
            if (embeddedImages.length > 0) {
                fetch(embeddedImages[0].src)
                    .then((response) => {
                        if (!response.ok) {
                            throw Error(response.statusText);
                        }
                        return response.blob();
                    })
                    .then((x) => {
                        const extension = IMAGE_EXTENSIONS[x.type];
                        if (!extension) {
                            throw new Error(
                                this.props.translate(
                                    'textInputFocusable.noExtentionFoundForMimeType',
                                ),
                            );
                        }

                        return new File([x], `pasted_image.${extension}`, {});
                    })
                    .then(this.props.onPasteFile)
                    .catch(() => {
                        const errorDesc = this.props.translate(
                            'textInputFocusable.problemGettingImageYouPasted',
                        );
                        Growl.error(errorDesc);

                        /*
             * Since we intercepted the user-triggered paste event to check for attachments,
             * we need to manually set the value and call the `onChangeText` handler.
             * Synthetically-triggered paste events do not affect the document's contents.
             * See https://developer.mozilla.org/en-US/docs/Web/API/Element/paste_event for more details.
             */
                        this.handlePastedHTML(pastedHTMLAsText);
                    });
                return;
            }

            this.handlePastedHTML(pastedHTMLAsText);
        }
    }

```

**Change this** 
https://github.com/Expensify/App/blob/36c6f8a2ccb4a846daff664036571c2cdf1fc164/src/components/TextInputFocusable/index.js#L242-L250

**To**

Strip off the  unwanted html tags. 

```
  handlePastedHTML(html) {
        const santizedText = html.replace(
            /<\/?(span|div|svg|input|path|a|...)\b[^<>]*>/g,
            '',
        );
        try {
            document.execCommand('insertText', false, santizedText);
            this.updateNumberOfLines();
            // eslint-disable-next-line no-empty
        } catch (e) {}
    }
```

**SCREEN RECORDING**

    

https://user-images.githubusercontent.com/4717301/155875274-e12f58b4-d934-4a93-b3b7-bfe7df2cbc84.mov

**Comment by parasharrajat** [comment]
Thanks for the proposal @dennismunene. Did you try?
```
 `<someComponent />` 
 ````
  as message.

**Comment by dennismunene** [comment]
@parasharrajat  yes I have , works as expected.

**Comment by parasharrajat** [comment]
Ok. Thanks but I don't consider this a proper approach. 
```js
const parser = new ExpensiMark(); 
     const markdownText = parser.htmlToMarkdown(html); 
```

This step is important for the app.

**Comment by dennismunene** [comment]
@parasharrajat updated my proposal to include this step
```
const parser = new ExpensiMark(); 
const markdownText = parser.htmlToMarkdown(html); 

```

**Comment by dennismunene** [comment]
@parasharrajat updated my proposal.

**Comment by parasharrajat** [comment]
Thanks, @dennismunene but I still think we can do better. 
1. StripHTML is doing its job of removing HTML but the question is why are getting HTML at this stage? HTML should have been converted to MD by now.

Now there are two things to compare.
1. Copy to clipboard and Selection copy via `Ctrl+C`. 

Copy to clipboard works. But the selection does not. So it must have given you the idea by now what is wrong.

**Comment by mateusbra** [comment]
> Thanks for the proposal @mateusbra. Could you please root cause details to your proposal? and tell us how did you reach your solution.
@parasharrajat thank you for the review. I think this is gonna be more complex than I thought, please ignore my proposal by now

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the Exported assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @adelekennedy is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcaaron (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by puneetlath** [comment]
@marcaaron I'm about to head out on my two-month sabbatical, so handing this off to you. Thanks!

**Comment by marcaaron** [comment]
Still waiting for a proposal.

**Comment by adelekennedy** [comment]
Doubling the issue!

**Comment by adelekennedy** [comment]
Not overdue - just doubled the issue 4 days ago

**Comment by MelvinBot** [comment]
@parasharrajat, @marcaaron, @adelekennedy Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by marcaaron** [comment]
🦗

**Comment by adelekennedy** [comment]
Still no volunteers 😬 we're a day out from me doubling the issue

**Comment by adelekennedy** [comment]
🤑 doubled 🤑

**Comment by marcaaron** [comment]
Melv needs a spa day

**Comment by Beamanator** [comment]
Maybe this should be `weekly`?

**Comment by parasharrajat** [comment]
In case someone is looking to solve this, https://github.com/Expensify/App/issues/7912#issuecomment-1053609946 indicates what needs to be done.

**Comment by adelekennedy** [comment]
We're at the mark to double the issue again! @parasharrajat you've pointed out how this issue can be resolved above @mateusbra and @dennismunene any interest in giving a proposal another go?

**Comment by Big-Vi** [comment]
Hi,

I just looked into this issue now.  As mentioned in the previous comment, stripHTML function from expensify-common library strips out the text within <>.

[](https://github.com/Expensify/expensify-common/blob/main/lib/str.js#L312)

Modifying the regex works fine. The video is attached below:

https://user-images.githubusercontent.com/30820063/160226276-a10e4167-b4c2-4c33-b0aa-cb42bb0e6e49.mov

I know the above solution works. but what if we take this opportunity to build a text formatting toolbar like slack(Video attached below):

https://user-images.githubusercontent.com/30820063/160227053-a9a44475-01f9-45e7-9df6-2108b7bd3c32.mov

With this approach basically, We would replace the <textarea> input with Richtoolbar.

**Comment by barun1997** [comment]
Hey, I just looked into the above issue, and here are my thoughts:

The current bug originates when we are converting a clean final HTML to Markdown (for instance,this is what the htmlToMarkdown helper takes: `<p>This is component <Component /> </p>`)

Within the parser, it processes the HTML string through the given markdown rules. 

The given markdown rules don't consider out-of-scope HTML Tags, which is why it's still returning HTML. 

To address this comment:  https://github.com/Expensify/App/issues/7912#issuecomment-1053609946, my proposal is to handle this within `HTMLToMarkdown` itself. So this would be to add one more rule in `htmlToMarkdownRules` in `ExpensiMark`. (https://github.com/Expensify/expensify-common/blob/f77bb4710c13d01153716df7fb087b637ba3b8bd/lib/ExpensiMark.js#L174-L190) 

This rule will replace all valid HTML tags but one’s we don’t have a rule for with empty string (as suggested here before).

```
                    {
                                    name: 'all-other-tags',
                                    regex: /<\/?(a|abbr|acronym|address|applet|area|article|aside|audio|b|base|basefont|bdi|bdo|bgsound|big|blink|blockquote|body|br|button|canvas|caption|center|cite|code|col|colgroup|data|datalist|dd|del|details|dfn|dir|div|dl|dt|em|embed|fieldset|figcaption|figure|font|footer|form|frame|frameset|h1|h2|h3|h4|h5|h6|head|header|hgroup|hr|html|i|iframe|img|input|ins|isindex|kbd|keygen|label|legend|li|link|listing|main|map|mark|marquee|menu|menuitem|meta|meter|nav|nobr|noframes|noscript|object|ol|optgroup|option|output|p|param|plaintext|pre|progress|q|rp|rt|ruby|s|samp|script|section|select|small|source|spacer|span|strike|strong|style|sub|summary|sup|table|tbody|td|textarea|tfoot|th|thead|time|title|tr|track|tt|u|ul|var|video|wbr|xmp|path|svg|...)\b[^<>]*>/g,
                                    replacement: '',
                    }
```

And we can remove Str.stripHTML from htmlToMarkdown helper.

**Comment by brianmuks** [comment]
***Proposal***
The issue is coming from this line :
https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/libs/SelectionScraper/index.js#L135
Wrong html content is being passed to `parser.htmlToMarkdown(newHtml);`
If you compare with what is being passed to :
https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/pages/home/report/ContextMenu/ContextMenuActions.js#L69

When a user clicks on `copy to cliboard` for this code block : "`<someComponent />`" what is being passed to  `parser.htmlToMarkdown(newHtml);` is "<code>&lt;someComponent /&gt;</code>". This is similar to the data coming from :
 https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/libs/actions/Report.js#L946
 
 ` html: '<code>&lt;someComponent /&gt;</code>'`
 
 With that said in order to pass the right format to `parser.htmlToMarkdown(newHtml);` when a user presses `CTR+C` we need to know what the original text looked like . To achieve this we need a new function to generate the text:
  
 ```

const getTextFromSelection = (selectionHtml) => {

    const div = document.createElement('div');
    div.innerHTML = selectionHtml;
    //get the last child from parent node 
    let lastNode = div.lastChild;
    let innerHTML = '';

   //loop through the node until we get the content of the lastNode which has the original text

    while (lastNode && lastNode.hasChildNodes) {
        innerHTML = lastNode.innerHTML ? lastNode.innerHTML : innerHTML;
        lastNode = lastNode.lastChild;

    }
    return innerHTML;
}
```
The results of the  `getTextFromSelection` is " &lt;someComponent /&gt;"
So all we are missing here is just the wrapper which can be obtained from the results of :
https://github.com/Expensify/App/blob/b347866b2af95bd2c51eb6bb65098bd1d756e7d0/src/libs/SelectionScraper/index.js#L131
the results `render(domRepresentation);` is "<code><span><someComponent /></span></code>"
At this point we just need to replace "<someComponent />" with the results from `getTextFromSelection` .
` const newHtmlWithText = newHtml.replace(selection, originalText);`
Now we can parse newHtmlWithText to `parser.htmlToMarkdown(newHtmlWithText);`.

The modification of `getAsMarkdown` will be :

```
const getAsMarkdown = () => {
    const selectionHtml = getHTMLOfSelection();
    const selection = window.getSelection();

    const domRepresentation = parseDocument(selectionHtml);

    domRepresentation.children = _.map(domRepresentation.children, c => replaceNodes(c));

    const newHtml = render(domRepresentation);

    const originalText = getTextFromSelection(selectionHtml);
    console.log(':domRepresentation', newHtml);

    //replace the the rendered text with the original text
    const newHtmlWithText = newHtml.replace(selection, originalText);

    const parser = new ExpensiMark();

    return parser.htmlToMarkdown(newHtmlWithText);
};
```



  

https://user-images.githubusercontent.com/4520893/160298441-5eb4cd87-012d-4d41-8441-e38534a5eebf.mov

**Comment by brianmuks** [comment]
> Edit: @brianmuks seems to have nailed it below. So retracting my proposal.

Thanks @barun1997

**Comment by parasharrajat** [comment]
@barun1997 Thanks for the proposal. Could be please test your proposal against the following: 

Paste this in composer and send as a message.
```
> This is a quote
> Mul

*Bold* _italic_ https://google.com

`yo yo code block` *bold inline*

```

On which fact are you proposing the string replacement as a fix? What is the base of this solution? 

 ```
 
    const originalText = getTextFromSelection(selectionHtml);
    console.log(':domRepresentation', newHtml);

    //replace the the rendered text with the original text
    const newHtmlWithText = newHtml.replace(selection, originalText);
  ```
  
  Why does doing this solve the problem? Will it work in general with every tag?

**Comment by brianmuks** [comment]
My solution does not cover these scenario:
Sentence with <code/> in it
So I though of some minor modifications:

```
const getAsMarkdown = () => {
    const selectionHtml = getHTMLOfSelection();
    const originalText = getTextFromSelection(selectionHtml);


    // const domRepresentation = parseDocument(selectionHtml);
    // domRepresentation.children = _.map(domRepresentation.children, c => replaceNodes(c));

    // const newHtml = render(domRepresentation);

    const parser = new ExpensiMark();

    return parser.htmlToMarkdown(originalText);
};
```
```
const getTextFromSelection = (selectionHtml) => {

    const div = document.createElement('div');
    div.innerHTML = selectionHtml;
    let node = div;
    let innerHTML = '';
    let counter = 0;

    //loop through the node until we get the content of the lastNode which has the original text

    while (node && node.hasChildNodes) {

        //ignore the content of first child as experiment shows that the main parent's 
        //content is equal to the first child's content
        if (node && node.children && node.children.length && counter !== 1) {
            // console.log('nodeChildrenHasChild:' + node.children);

            var children = [].slice.call(node.children);
            children.forEach(child => {

                const container = { ...child.dataset }.testid;
                //if child is a code component wrap it in its original component
                innerHTML += container ? `<${container}>${child.firstChild.innerHTML}</${container}>` : child.innerHTML;

            });

        }
        node = node.lastChild;
        counter++;
    }

    return innerHTML;
}
```

**Comment by parasharrajat** [comment]
Could you please help with answers to my questions as well? Also, small(any) details will help me understand your plan. https://github.com/Expensify/App/issues/7912#issuecomment-1080342870 does not tell what are you trying to establish.

The previous proposal suggests that you are proposing string replacement as a solution and in the new code snippet, your function `getTextFromSelection` is returning HTML. Why?


Although you are looking at the correct place it seems you may have been distracted from the main solution here. I will suggest finding another simpler approach to the problem.

**Comment by brianmuks** [comment]
> Could you please help with answers to my questions as well? Also, small(any) details will help me understand your plan. [#7912 (comment)](https://github.com/Expensify/App/issues/7912#issuecomment-1080342870) does not tell what are you trying to establish.
> 
> The previous proposal suggests that you are proposing string replacement as a solution and in the new code snippet, your function `getTextFromSelection` is returning HTML. Why?
> 
> Although you are looking at the correct place it seems you may have been distracted from the main solution here. I will suggest finding another simpler approach to the problem.

So what we are trying to achieve here is to extract/determine the original text that was saved to the database as `html`. 
The `copy to clipboard`   function basically  uses the `html`  content and passes it to `parser.htmlToMarkdown(html);` then writes it to the clipboard. So I feel that the function handling `CTR+C`  should be as similar as possible to the 
`copy to clipboard` function. 

If there was an easy way of extracting/determining the original `html` fetched from the database for a message from `getHTMLOfSelection();` then we could use that.  

By the way `copy to clipboard` fails to copy quotes properly. So I was wondering if we want to fix that all at once because I reported a bug for it few days ago.





https://user-images.githubusercontent.com/4520893/160365106-2f28baf3-49b4-4657-90a2-f9c7fa1d437a.mov

So far  my latest solution works for tags . I am still looking at how to handle  scenarios you asked me to test. 
`"> This is a quote"`
`"> Mul"`

`"*Bold* _italic_ https://google.com"`

`"`yo yo code block` *bold inline*"`

https://user-images.githubusercontent.com/4520893/160367877-cfbed48f-db44-4a52-b28d-9a74adaea9e2.mov

**Comment by parasharrajat** [comment]
> So what we are trying to achieve here is to extract/determine the original text that was saved to the database as html.

Thanks for the explanation. It would be technically impossible with the current implementation of message parsing. Also, it does not seem like an elegant solution. 

> text += container ? `<${container}>${child.innerHTML}</${container}>` : child.innerHTML;

It does sound like HTML.

> By the way copy to clipboard fails to copy quotes properly. So I was wondering if we want to fix that all at once because I reported a bug for it a few days ago.

we can only focus on solving the code blocks in a way that does not block or break others.

**Comment by gladiator-1** [comment]
### Proposal

in src/libs/SelectionScraper :
```
const replaceNodes = (dom) => {
+   if (dom.type == "text") {
+        dom.data = dom.data.replaceAll('<', '&lt;').replaceAll('>', '&gt;')
+   }
    let domName = dom.name;
```

Result : 

https://user-images.githubusercontent.com/102553449/160479316-26a46b18-9ebf-4ead-aa12-bf6d79901db7.mp4

**Comment by parasharrajat** [comment]
Thanks for the proposal @gladiator-1. It seems to do what we need here. Could you please confirm if we can use `Str.htmlEncode` here instead of manual replacement?

**Comment by gladiator-1** [comment]
@parasharrajat Sure

```
  import {parseDocument} from 'htmlparser2';
  import {Element} from 'domhandler';
  import _ from 'underscore';
+ import Str from 'expensify-common/lib/str';




const replaceNodes = (dom) => {
+   if (dom.type == "text") {
+       dom.data = Str.htmlEncode(dom.data)
+   }
    let domName = dom.name;
```

**Comment by parasharrajat** [comment]
> Could you please confirm if we can use `Str.htmlEncode` here instead of a manual replacement?

Is it working fine without breaking other tags?

**Comment by gladiator-1** [comment]
>Is it working fine without breaking other tags?

Yes

https://user-images.githubusercontent.com/102553449/160613681-4475a20a-f809-4f7a-a8fc-76789534417c.mp4

**Comment by parasharrajat** [comment]
Thanks, I was really looking for a simpler approach like @gladiator-1 has presented.  

One improvement would be to do this manipulation at the parsing step [onText](https://github.com/fb55/htmlparser2#usage). 


Thanks, @barun1997 and @brianmuks for your efforts in looking into getting a proper solution. But I like @gladiator-1's simpler approach to the problem. https://github.com/Expensify/App/issues/7912#issuecomment-1081094208 https://github.com/Expensify/App/issues/7912#issuecomment-1081733070

cc: @marcaaron 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by gladiator-1** [comment]
>One improvement would be to do this manipulation at the parsing step [onText](https://github.com/fb55/htmlparser2#usage).

Excuse me, Is it possible to update the parsed document object at the parsing step ?

 `const domRepresentation = parseDocument(selectionHtml)`

**Comment by parasharrajat** [comment]
I looked at the wrong name. It seems not, then your proposal is fine.

**Comment by adelekennedy** [comment]
@gladiator-1 I'm struggling to find you in Upwork, do you mind applying for the job [here](https://www.upwork.com/jobs/~01c90fb0c06dde1e5c)?

**Comment by gladiator-1** [comment]
@adelekennedy applied with code : 75340

**Comment by MelvinBot** [comment]
📣 @gladiator-1 You have been assigned to this job by @adelekennedy!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by gladiator-1** [comment]
@parasharrajat PR ready https://github.com/Expensify/App/pull/8449

**Comment by marcaaron** [comment]
This is in review now.

**Comment by gladiator-1** [comment]
@parasharrajat  I'm so sorry for this error, PR https://github.com/Expensify/App/pull/8741

**Comment by adelekennedy** [comment]
@parasharrajat do we still need to wait on payment or are we good to go with the fix in #8741?

**Comment by parasharrajat** [comment]
Yes, we have to wait. The new PR is not merged.

**Comment by adelekennedy** [comment]
it's okay Melvin - we're waiting on the new PR

**Comment by marcaaron** [comment]
Unassigning as I'm going OOO for a week. I think we are pretty much good here and just need to assign a new reviewer with PullerBear for whatever PR we are waiting on.

**Comment by adelekennedy** [comment]
checking here - it looks like #8741 has been merged, are we good for payment @parasharrajat?

**Comment by adelekennedy** [comment]
paid and paid!
